### ISSUE TITLE
Exploring Session Context using Distributed Representations of Queries and Reformulations

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/deep-learning-readinglist#103]
https://www.microsoft.com/en-us/research/wp-content/uploads/2015/08/sigirfp093-mitra.pdf

**Comment by sweaterr** [comment]
검색 로그에는 자주 발생하는 사용자 질의 재구성 패턴의 예가 들어 있습니다. 직관적으로, 재구성 "샌프란시스코"→ "샌프란시스코 49ers"는 의미 적으로 "디트로이트"→ "디트로이트 라이온스"와 유사합니다. 마찬가지로 "런던"→ "런던에서해야 할 일"과 "뉴욕"→ "뉴욕 관광 명소"도 비슷한 전환으로 간주 할 수 있습니다. 그러나 "영화"→ "새 영화"와 "뉴욕"→ "뉴욕"은 두 개의 재구성에서 어휘 유사점에도 불구하고 분명히 다르다. 본 논문에서는 Convolutional Latent Semantic Model과 같은 심 신경 네트워크 모델에 의해 학습 된 질의의 분산 표현을 연구하고 질의 재구성을 벡터로 표현하는 데 사용될 수 있음을 보여준다. 이러한 재구성 벡터는 의미적으로 및 구문론적으로 유사한 쿼리 변경을 매립 공간에 더 가깝게 매핑하는 것과 같은 유리한 특성을 나타낸다.

**Comment by sweaterr** [comment]
Furthermore, we show that the distributed representations of queries and reformulations are both useful for modelling session context for query prediction tasks, such as for query auto-completion (QAC) ranking. Our empirical study demonstrates that short-term (session) history context features based on these two representations improves the mean reciprocal rank (MRR) for the QAC ranking task by more than 10% over a supervised ranker baseline. Our results also show that by using features based on both these representations together we achieve a better performance, than either of them individually.
한국어
또한, 쿼리 및 재구성의 분산 된 표현은 쿼리 자동 완성 (QAC) 순위 지정과 같은 쿼리 예측 작업에 대한 세션 컨텍스트를 모델링하는 데 유용하다는 것을 보여줍니다. 우리의 경험적 연구는이 두 표현에 기반한 단기 (세션) 내역 컨텍스트 특성이 감독 된 순위 기반 기준보다 10 % 이상 QAC 순위 태스크의 평균 상호 순위 (MRR)를 향상 시킨다는 것을 입증합니다. 우리의 결과는 또한이 두 표현을 기반으로 한 피쳐를 함께 사용함으로써 어느 하나의 피사체보다 더 나은 성능을 얻을 수 있음을 보여줍니다.

**Comment by sweaterr** [comment]
In text processing, Mikolov et al. [31] demonstrated that the distributed representation of words learnt by continuous space language models are surprisingly good at capturing syntactic and semantic relationships between the words. Simple algebraic operations on the word vectors have been shown to produce intuitive results. For example, vector(“king”)−vector(“man”)+vector(“woman”) results in a vector that is in close proximity to the vector(“queen”). In Section 3, we will show that the embeddings learnt by the Convolutional Latent Semantic Model (CLSM) [36] exhibit similar favourable properties and hence provide an intuitive mechanism to represent query reformulations as the offsets between the query vectors

**Comment by sweaterr** [comment]
4 절에서 설명 된 우리의 경험적 연구는 질의와 재구성의 벡터 표현이 질의 제안 검색을위한 세션 컨텍스트를 캡처하는 데 유용 할 수 있음을 보여줍니다. CLSM 모델은 비슷한 의도를 지닌 쿼리 (및 문서)를 의미 공간의 동일한 이웃으로 매핑하도록 훈련되었습니다. 따라서 후보 제안과 사용자의 최근 질의 사이의 국소 적 유사성을 측정하는 데 적합합니다. 또한 우리의 실험은 사용자의 이전 검색어에서 후보 제안까지의 재구성의 벡터 표현이 제안의 관련성을 예측하는 데 유용한 신호가 될 수 있음을 보여줍니다. 우리는이 벡터 표현에 기반한 세션 컨텍스트 기능이 Shokouhi [37]에 의해 제안 된 감독 된 순위 기준보다 QAC 순위를 크게 향상시킬 수 있음을 보여주는 섹션 5에서 결과를 제시합니다.

**Comment by sweaterr** [comment]
3. MODELS
Architecture. We adopt the CLSM architecture proposed by Shen et al. [36] for our study. As illustrated in Figure 1, the CLSM model is a deep neural network with a convolutional-pooling structure that projects a variable-length text into a fixed-length real-valued vector. Each word in the input text (e.g., query) is first hashed to a letter trigram vector as described by Huang et al. [21]. The word hashing step, for example, maps the word "top" to a feature vector with nonzero values corresponding to the trigrams "#to", "top" and "op#", where "#" denotes the word boundary. It has been pointed out in previous work [21] that the trigram hashing technique is a robust strategy for dealing with misspellings and morphological variants of words in the data. It also scales much better for large vocabularies compared to the one-hot2 representation where each unique word is associated with a separate identifier. The number of unique words in a Web corpus can be extremely large, whereas the number of distinct letter trigrams tends to be much more manageable (limited to the top 50K for this study). These factors makes the trigram hashing strategy particularly useful for Web search.

**Comment by sweaterr** [comment]
Next, for each word the convolutional layer extracts contextual features based on its immediate neighbours as defined by a predetermined window size. A max pooling layer combines the output of the convolutional layer into a fixed-length feature vector. The max pooling layer in turn is connected to the output layer which produces the final vector representation of the whole query. Unless specified otherwise, for all models in this paper the window size for the convolutional layer is set to three and the dimensions of the output vector to 32.

**Comment by sweaterr** [comment]
**Training.** The training data for the CLSM models consists of source-target text pairs. The original DSSM [21] and CLSM [36] models were trained on clickthrough data which consists of pairs of queries and document titles, corresponding to clicked results.

**Comment by sweaterr** [comment]
In addition to clickthrough data, we also train the CLSM models on sampled pairs of queries from search logs that were observed in succession during user sessions. In the rest of this paper, we refer to this as the session pairs dataset. For a pair of observed queries {q1, q2}, if the dataset includes both the ordering {q1, q2} and {q2, q1} then we refer to it as the symmetric session pairs dataset, otherwise as asymmetric. The symmetric session pairs data is further randomly sub-sampled by half to keep the count of the training pairs in both the datasets comparable.

**Comment by sweaterr** [comment]
The session pairs datasets are extracted from the exact same user sessions from which the clickthrough data is generated. While this does not imply that the actual count of training pairs in these two types of datasets are equal, it does make the comparison more meaningful as it assumes the same amount of raw log data is examined for training both the types of models. In practice, however, we did observe the data sizes to be comparable across all three datasets during this study.

**Comment by sweaterr** [comment]
**Distributed Representations.** The CLSM models project the queries to an embedding space with fixed number of dimensions. The semantic similarity between two queries q1 and q2 in this semantic space is defined by,
![image](https://user-images.githubusercontent.com/1518919/27009424-0daab7d0-4ec9-11e7-95b2-6e6eee447e2b.png)
where y1 and y2 are the CLSM vector outputs corresponding to the two queries, respectively. A close examination of the CLSM output vectors reveal that the learnt distributed representations hold useful information about inter-query relationships. Figure 2 illustrates how the offset vectors between pairs of queries, represented by the directed edges, are directionally similar in the embedding space for similar intent transitions. This matches the observations made by Mikolov et al. [29] on continuous space language models for text processing, and gives us an intuitively understandable representation of query reformulations as their offset vectors in the embedding space. More specifically, we define the reformulation from query q1 to q2 as,

**Comment by sweaterr** [comment]
More specifically, we define the reformulation from query q1 to q2 as,

![image](https://user-images.githubusercontent.com/1518919/27009622-baeacfee-4ecd-11e7-9a93-d68ac010344d.png)

where y1 and y2 are the CLSM vector embeddings of the two queries, respectively. This explicit vector representation provides a framework for studying frequently occurring query reformulation 
patterns. To illustrate this, we randomly sample approximately 65K pairs of queries that were observed in succession in Bing’s logs. For each pair, we compute the offset vector using a CLSM model. We then run a simple k-means clustering (k = 100) and examine the top clusters. Example reformulations from five of the biggest clusters are shown in Table 1.

**Comment by sweaterr** [comment]
A further study of these reformulation vectors can reveal important insights about user behaviour, such as the popularity of certain reformulation patterns. For example, we randomly sampled 100,000 adjacent pairs of queries from Bing’s logs that were observed in search sessions. Our analysis show that there are more pairs similar to the narrowing reformulation "new york" → "things to do in new york" in the sampled set, than its inverse. Similarly, the misspelling "fcebook" followed by "facebook" is a more commonly observed pattern than the other way around, as illustrated in Figure 3.

**Comment by sweaterr** [comment]
4. EXPERIMENT SETUP
Our empirical evaluations are based on the learning to rank framework proposed by Shokouhi [37] for personalized query autocompletions. In this setup, we learn a supervised ranking model based on training data generated from implicit user feedback. The output of the CLSM models, described in the previous section, are used to generate additional features for this supervised ranking model. The baseline ranking model (henceforth referred to simply as the baseline model) contains both the non-contextual and the (non-CLSM based) contextual features. We compare all models using the mean reciprocal rank (MRR) metric, and the study is repeated on two different testbeds to further confirm the validity of the results.

**Comment by sweaterr** [comment]
4.1 Testbeds

We conduct our experiments on a large scale search query dataset sampled from the logs of the Bing search engine. We also reproduce our results using the publicly available AOL query logs [33]. In the rest of this paper we refer to these two datasets as the Bing testbed and the AOL testbed, respectively.

**Comment by sweaterr** [comment]
Bing testbed. 

Bing’s logs contain a record of all the queries submitted by its users associated with the corresponding anonymized user IDs, timestamps and any clicked Web results3 (the URL and the displayed title). We sampled queries from these logs for the duration of the last week of October, 2014 and use this as the background data, for computing the feature values and training the CLSM models. From the first week of November, we sampled 175,392 queries from two consecutive days for training the supervised ranking models, and from the following two individual days we sampled 79,000 queries for validation and 74,663 queries for testing, respectively.

**Comment by sweaterr** [comment]
4.2 Learning to rank

To generate the training, the validation and the test sets we sample query impressions from the corresponding portions of the logs. For each query impression, a prefix is generated by splitting the query at a randomly selected position4 . For each prefix a positive relevance judgment is assigned to the suggestion candidate that matches the final submitted query and all the others are labelled as irrelevant.

**Comment by sweaterr** [comment]
The training data collected in the above process consists of labelled prefix-query pairs. With respect to the choice of learningto-rank algorithms, we chose LambdaMART [42], a boosted tree version of LambdaRank [6], that won the Yahoo! Learning to Rank Challenge (2010) [9] and is considered as one of the state-of-the-art learning algorithms. We train 500 trees across all our experiments with the same set of fixed parameters tuned using standard training and validation on separate sets.

**Comment by sweaterr** [comment]
We consider the top 10 million most popular queries in the background data as the pre-computed list of suggestion candidates and filter out all the impressions where the final submitted query is not present in this list. For each impression in the training, the validation and the test sets we retain a maximum of 20 suggestion candidates - the submitted query as the positive candidate and 19 other most frequently observed queries from the background data that starts with the same prefix, as the negative examples. Furthermore, for each impression up to 10 previous queries from the same session are made available for computing the session context features. Similar to other previous work [11, 22] we define the end of a session by a 30 minute window of user inactivity.

**Comment by sweaterr** [comment]
For our final evaluation we report the Mean Reciprocal Rank of the submitted query averaged over all sampled impressions on each of the two testbeds.

**Comment by sweaterr** [comment]
4.3 Features

Non-contextual features. 

The MostPopularCompletion (MPC) model is one of the baselines for our study. We also use the output of this model as a feature for the supervised ranking model. Other non-contextual features include the prefix length (in characters), the suggestion length (in both characters and words), the vowels to alphabets ratio in the suggestion and a boolean feature indicating whether the suggestion contains numeric characters.

N-gram similarity features. 

We compute the character ngram similarity (n=3) between the suggestion candidate and the
previous queries from the same user session. This is an implementation of the short history features described by Shokouhi [37]. A maximum of 10 previous queries are considered.

Pairwise frequency feature. 

From the background data, we generate the top 10 million most popular adjacent pairs of queries
observed in search sessions.  For a given impression, the previous query and the suggestion candidate pair is matched against this dataset and the corresponding frequency count is used as the feature value.  If no matches are found, then the feature value is set to zero.

CLSM topical similarity features. 

The CLSM models are trained as described in Section 3 using the background portion of
the data on each testbed. The cosine similarity between the CLSM vectors corresponding to the suggestion candidate and a maximum of previous 10 queries from the same session are computed and used as 10 distinct features in the QAC ranking model.

Training on the session query pairs data produces a pair of prepost CLSM models. When trained on the asymmetric data, the premodel is used for projecting the user’s previous queries and the postmodel is used for projecting the suggestion candidates for the cosine similarity computation. For the symmetric data however, both the pre- and the post- models are equivalent, and hence we use only the pre- model in our experiments. 

The AOL logs contains only the domain portion of the clicked results. Hence we are unable to get the corresponding document titles. Therefore we only train the session pairs based CLSM models on this testbed and report those results in this paper.

CLSM reformulation features. 

We compute the n-dimensional (n=32) vector representation of the reformulation from the previous query to the suggestion candidate. The raw values from this vector are used as n distinct features into the supervised ranking model. For both the session pair based models, the pre- model is used for projecting the suggestion candidates, as well as the previous query.

**Comment by sweaterr** [comment]
5. RESULTS

Table 3 compares the results of training the supervised QAC ranking model with the different CLSM based session context features. Due to the proprietary nature of Bing’s data, we report only relative
improvements of each of the models over the MPC baseline for this testbed. On the AOL testbed, however, we report both the  absolute MRR values and the relative improvements for all the models.

![image](https://user-images.githubusercontent.com/1518919/27010631-dc805abe-4ee3-11e7-86f3-b93b72c5b5e7.png)

**Comment by sweaterr** [comment]
On both the testbeds, the baseline model which also contains
session context features (the n-gram similarity and the pairwise frequency)
shows a large improvement over the MPC baseline, which
is expected. All the models trained with the CLSM based contextual
features show further statistically significant improvements over the
baseline model. Both the CLSM models trained on session pairs
perform better than the models trained on clickthrough data, with
the model trained on the symmetric session pairs performing slightly
better overall. Table 5 lists examples of cases from one of the test
sets where the ranking model with the CLSM based contextual
features perform better compared to both the baselines.

**Comment by sweaterr** [comment]
The supervised ranking models trained with both the CLSM based similarity features and the CLSM based reformulation features perform better than the corresponding models trained with the similarity features alone. The improvements are statistically significant and demonstrate the additional information provided by the reformulation features to the ranking model over the CLSM based similarity features. The reformulation features perform particularly superior when the CLSM model has been trained on the session pairs dataset

**Comment by sweaterr** [comment]
Table 4 shows the impact of considering different number of previous queries in the session for computing the CLSM based similarity features. The results indicate that considering the previous
query alone achieves most of the improvements observed from these similarity features.

![image](https://user-images.githubusercontent.com/1518919/27010812-8cdc8000-4ee8-11e7-9923-7a5dbebec3f0.png)

**Comment by sweaterr** [comment]
We also compare the improvements from the different models
based on the length of the input prefixes. Bar-Yossef and Kraus [1]
have previously reported that non-contextual QAC systems generally
perform poorly when the user has typed only a few characters
due to the obvious ambiguity in user intent. Figure 4 illustrates this
behaviour on the AOL testbed. Both the supervised ranking models,
the baseline and the model with the CLSM features, show signifi-
cantly large improvements over the MPC baseline on short prefixes.
After the user has typed a few more characters in the search box,
the set of suggestion candidates reduce significantly and the performance
of the MPC model improves. Therefore the improvements
on the longer prefixes are smaller for both the supervised ranking
models. The supervised ranking model with the CLSM features,
however, show statistically significant better MRR compared to both
the MPC baseline and the supervised baseline ranking model on all
the prefix length based segments. Finally, Figure 5 shows that better
MRR can be achieved by training the CLSM model with a higher
number of output dimensions.

**Comment by sweaterr** [comment]
6. DISCUSSION

In the previous section we demonstrated significant improvements in the query auto-completion ranking task using the CLSM based session context features. We now discuss potential implications of these vector representations on session modelling and list some of the assumptions and limitations of the evaluation framework used in this study.

**Comment by sweaterr** [comment]
Implications for session modelling. 

The distributed representation of queries and query reformulations provides an interesting
framework for thinking about sessions and task context. The sequence of queries (and documents) in a search session can be
considered as a directed path in the embedding space. What are the
common attributes shared by these session paths? What properties
of these paths vary depending on the type of the user task or information
need? These are examples of research questions that may
be interesting to study under the distributed representation framework.
Hassan et al. [16], for example, studied long search sessions
and compared user behaviours when the user is struggling in their

**Comment by sweaterr** [comment]
In this paper we have examined individual query reformulations.
Studying reformulation chains may teach us further about how user
intents evolve during a session and support the design of future
models for session search. For example, White and Huang [40]
have explored the value of search trails, over the origins and the
destinations. While we have only examined the representation of
queries and reformulations in this paper, CLSM also allows for
documents to be represented in the same embedding space. A
unified study of queries, reformulations and viewed (searched or
browsed) documents using the vector representation framework is
an area for future work.

**Comment by sweaterr** [comment]
Assumptions and limitations. We have based our empirical study on the supervised ranking framework proposed by Shokouhi [37]. 
In doing so, we inherit some of the assumptions in the designs of that framework. Firstly, we assume that the user has a predetermined
query in mind for input and would be satisfied if it
appears in the QAC suggestions list. However Hofmann et al. [17]
have shown that due to the high examination bias towards top-ranked
results, sub-optimal QAC ranking can negatively affect the quality
of the query submitted by the user. As most popular Web search
engines implement some form of an auto-completion feature, it is
likely that those QAC systems influenced the actual query observed
in the logs. We ignore this effect in the generation of our training
and test sets.

**Comment by sweaterr** [comment]
7. CONCLUSION

We have demonstrated that the distributed representation of queries by the convolutional latent semantic models holds useful information about inter-query relationships. 

the reformulation vectors provide an interesting framework for studying sessions and intent progressions
